chown/chown.1
.\"
.\" Sccsid @(#)chown.1	1.9 (gritter) 1/29/05
.\" Parts taken from chown(1), Unix 7th edition:
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\"
.\" SPDX-Licence-Identifier: Caldera
.TH CHOWN 1 "1/29/05" "Heirloom Toolchest" "User Commands"
.SH NAME
chown, chgrp \- change owner or group
.SH SYNOPSIS
\fBchown\fR [\fB\-hHLPR\fR] \fIowner\fR[:\fIgroup\fR] \fIfile\fR ...
.PP
\fBchgrp\fR [\fB\-hHLPR\fR] \fIgroup \fIfile\fR ...
.SH DESCRIPTION
.B Chown
changes the owner of the
.I files
to
.IR owner ,
and, if
.I group
is present,
the group-ID of the
.I files
to
.IR group .
Both
.I owner
and
.I group
may be either a decimal
.SM UID
/
.SM GID
or a login / group name
found in the password file.
.PP
.B Chgrp
changes the group-ID of the
.I files
to
.IR group .
The
.I group
may be either a decimal
.SM GID
or a group name found in the group-ID file.
.PP
Both
.B chown
and
.B chgrp
accept the following options:
.TP
.B \-h
When a symbolic link is encountered,
change the owner and group of the link itself,
rather than the file it refers to.
.TP
.B \-R
Recursively descend into each
.I file
operand.
Whether owner and group of a symbolic link are changed
depends on the
.B \-h
option,
but the link is not followed otherwise.
.PP
The following options have been introduced by POSIX.1-2001:
.TP
.B \-H
With the
.I \-R
option, if a symbolic link given on the command line
points to a directory,
follow that symbolic link
and change owner and group in the files below,
but do not handle any other symbolic links specially.
.TP
.B \-L
With the
.I \-R
option, whenever a symbolic link is encountered
that points to a directory,
follow the symbolic link
and change owner and group in the files below,
.TP
.B \-P
With the
.I \-R
option, do not follow any symbolic links,
but change owner and group of the links themselves.
.SH FILES
.B /etc/passwd
.br
.B /etc/group
.SH "SEE ALSO"
chown(2),
passwd(5),
group(5)
.SH NOTES
Permission to change owner and group is based
either on historical System V behaviour,
which is to allow the owner of the file or a privileged user
a change to any value;
or on historical
.SM BSD
behaviour,
which is to restrict changing the owner
to a privileged user
and changing the group to a group to which the file owner belongs.
.PP
The [:\fIgroup\fR] argument to
.I chown
has been introduced with POSIX.2
and is not available on older implementations of System\ V.


CHANGES
Release ...
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants. 

Release 230503 
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin. 

Release 230120
* Notes for the library directories if linking against the GNU C library.
* rm: Consistency within standards: the SUS program shouldn't have the POSIX 1003.1(2008) fix, just the main program.
* Made $(ROOT) not hardcoded anymore.
* Cleaned up more licence headers (if my estimations are correct, we've got rid of 3.255 lines!).
* Small update on README.md --- although it's now less readable on plain-text readers such as more or pg.

Release 230107
* rm: Added support for the POSIX 1003.1(2008).
* chroot: Added manual page.
* awk: Added stdint.h to awk.g.y.
* Now it has a bigger range of compatibility with GNU C library versions.

Release 221025
* Made notes clearer about building Heirloom on modern versions of the GNU C library.
* tar: Support for zstd and xz.

Release 220821
* [NEW] chroot: New implementation of chroot, based on Solaris 2.5 payloads.
* true, false: Rewritten in C.
* yes: Rewritten and adapted to ANSI C.
* "libcommon/vpfmt.c": Adapted to ANSI C.
* Now many of the utilities can be built in the musl C library and GNU C library (dynamically).

Release 070715
* cpio: A "-P" option has been introduced that allows to specify attributes
  like the file owner when creating an archive. With it, unprivileged users
  can set arbitrary attributes for files within an archive.
* diff: Made it compile on CentOS 5.
* Static linking against glibc is now possible since "optopt" is no longer
  a common symbol, and there is no conflict anymore with "getopt.o".
* cpio, tar: A compilation problem on some Linux systems has been fixed.
* ul: A compilation problem for NetBSD and other systems that use termcap
  instead of curses has been fixed (patch by Aleksey Cheusov).
* bfs: An OpenSolaris-derived manual page is now included.

Release 070227
* The directory for development utilities, by default "/usr/ccs/bin", is
  now configurable in "build/mk.config" . It is expected that the utilities
  from the "Heirloom Development Tools" package are installed in it.
* getconf: "/usr/ccs/bin" is now included in the default path as printed
  with the "PATH" argument.
* what: Has been replaced by an OpenSolaris-derived implementation.
* tsort: Is now installed in "/usr/ccs/bin".
* OpenSolaris-derived manual pages are now included for bdiff, newform,
  and fspec(5).
* psrinfo: Fixed the computation of the "on-line since" time for Linux 2.6.

chown/chown.c
/*
 * chown - change file owner
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, July 2002.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * This software is provided 'as-is', without any express or implied
 * warranty. In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute
 * it freely, subject to the following restrictions:
 *
 * 1. The origin of this software must not be misrepresented; you must not
 *    claim that you wrote the original software. If you use this software
 *    in a product, an acknowledgment in the product documentation would be
 *    appreciated but is not required.
 *
 * 2. Altered source versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.
 *
 * 3. This notice may not be removed or altered from any source distribution.
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
#ifdef	UCB
static const char sccsid[] USED = "@(#)/usr/ucb/chown.sl	1.14 (gritter) 5/29/05";
#else
static const char sccsid[] USED = "@(#)chown.sl	1.14 (gritter) 5/29/05";
#endif

#include	<sys/types.h>
#include	<sys/stat.h>
#include	<unistd.h>
#include	<stdio.h>
#include	<string.h>
#include	<stdlib.h>
#include	<errno.h>
#include	<libgen.h>
#include	<limits.h>
#include	<stdarg.h>
#include	<dirent.h>
#include	<pwd.h>
#include	<grp.h>

#ifdef	UCB
#define	SEPCHAR	'.'
#else	/* !UCB */
#define	SEPCHAR	':'
#endif	/* !UCB */

static enum {
	PERS_CHOWN,
	PERS_CHGRP
} pers = PERS_CHOWN;

static unsigned	errcnt;			/* count of errors */
static int	fflag;			/* force (don't report errors) */
static int	hflag;			/* use lchown */
static int	Rflag;			/* recursive */
static int	HLPflag;		/* -H, -L, -P */
static char	*progname;		/* argv[0] to main() */
static char	*owner;
static char	*group;
static uid_t	myeuid;
static uid_t	new_uid;
static gid_t	new_gid;

static void *
srealloc(void *vp, size_t nbytes)
{
	void *p;

	if ((p = realloc(vp, nbytes)) == NULL) {
		write(2, "Out of memory\n", 14);
		exit(077);
	}
	return p;
}

static void *
smalloc(size_t nbytes)
{
	return srealloc(NULL, nbytes);
}

static int
eprintf(int always, const char *fmt, ...)
{
	va_list	ap;
	int	ret = 0;

	if (always || fflag == 0) {
		va_start(ap, fmt);
		ret = vfprintf(stderr, fmt, ap);
		va_end(ap);
	}
	errcnt |= 1;
	return ret;
}

static void
getpath(const char *path, char **file, char **filend, size_t *sz, size_t *slen)
{
	*sz = 14 + strlen(path) + 2;
	*file = smalloc(*sz);
	*filend = *file;
	if (path[0] == '/' && path[1] == '\0')
		*(*filend)++ = '/';
	else {
		register const char *cp = path;
		while ((*(*filend)++ = *cp++) != '\0');
		(*filend)[-1] = '/';
	}
	*slen = *filend - *file;
}

static void
setpath(const char *base, char **file, char **filend,
		size_t slen, size_t *sz, size_t *ss)
{
	if (slen + (*ss = strlen(base)) >= *sz) {
		*sz += slen + *ss + 15;
		*file = srealloc(*file, *sz);
		*filend = &(*file)[slen];
	}
	strcpy(*filend, base);
}

static void
dochown(const char *path, int level)
{
	struct stat ost, nst, lst;
	int	(*chownfn)(const char *, uid_t, gid_t);
	int	(*statfn)(const char *, struct stat *);

	chownfn = hflag || Rflag && HLPflag == 'P' ? lchown : chown;
	statfn = hflag || Rflag && HLPflag == 'P' ? lstat : stat;
	if (statfn(path, &ost) < 0) {
		eprintf(0, "%s: cannot access %s: %s\n",
				progname, path, strerror(errno));
		return;
	}
	if (chownfn(path, owner ? new_uid : ost.st_uid,
				group ? new_gid : ost.st_gid) < 0) {
		eprintf(0, "%s: cannot chown %s: %s\n",
				progname, path, strerror(errno));
		goto failed;
	}
	if ((ost.st_mode & S_ISUID) || (ost.st_mode & (S_ISGID|S_IXGRP))
				== (S_ISGID|S_IXGRP)) {
		if (statfn(path, &nst) < 0) {
			eprintf(0, "%s: cannot access %s after chmod: %s\n",
				progname, path, strerror(errno));
			return;
		}
		if (myeuid && ((nst.st_mode & S_ISUID) ||
					(nst.st_mode & (S_ISGID|S_IXGRP))
					== (S_ISGID|S_IXGRP))) {
			if (chmod(path, ost.st_mode & 07777 &
						~(S_ISUID|S_ISGID)) < 0) {
				eprintf(0,
				"%s: cannot clear suid/sgid bits of %s: %s\n",
					progname, path, strerror(errno));
			}
		} else if (myeuid == 0 && (ost.st_mode & 07777) !=
				(nst.st_mode & 07777)) {
			if (chmod(path, ost.st_mode & 07777) < 0) {
				eprintf(0, "%s: cannot reset modes of %s: %s\n",
					progname, path, strerror(errno));
			}
		}
	} else
	failed:	nst = ost;
	if (Rflag == 0)
		return;
	if (statfn != lstat && (HLPflag != 'H' || level > 0)
			&& HLPflag != 'L') {
		if (lstat(path, &lst) < 0) {
			eprintf(0, "%s: cannot lstat %s: %s\n",
					progname, path, strerror(errno));
			return;
		}
	} else
		lst = nst;
	if (S_ISDIR(lst.st_mode)) {
		DIR *Dp;
		struct dirent *dp;
		char *copy, *cend;
		size_t sz, slen, ss;

		if ((Dp = opendir(path)) == NULL) {
			eprintf(0, "%s: cannot open directory %s: %s\n",
					progname, path, strerror(errno));
			return;
		}
		getpath(path, &copy, &cend, &sz, &slen);
		while ((dp = readdir(Dp)) != NULL) {
			if (dp->d_name[0] == '.' && (dp->d_name[1] == '\0' ||
						(dp->d_name[1] == '.' &&
						 dp->d_name[2] == '\0')))
				continue;
			setpath(dp->d_name, &copy, &cend, slen, &sz, &ss);
			dochown(copy, level + 1);
		}
		free(copy);
		closedir(Dp);
	}
}

static long
numeric(const char *name, const char *type)
{
	char *x;
	long n;

	n = strtol(name, &x, 10);
	if (*x != '\0' || *name == '-' || *name == '+') {
		eprintf(1, "%s: unknown %s %s\n", progname, type, name);
		exit(2);
	}
	return n;
}

static void
getowner(char *name)
{
	if (pers == PERS_CHOWN) {
		owner = name;
		if ((group = strchr(name, SEPCHAR)) != 0)
			*group++ = '\0';
	} else if (pers == PERS_CHGRP)
		group = name;
	if (owner) {
		struct passwd *pwd;

		if ((pwd = getpwnam(owner)) == NULL)
			new_uid = numeric(owner, "user id");
		else
			new_uid = pwd->pw_uid;
	}
	if (group) {
		struct group *grp;

		if ((grp = getgrnam(group)) == NULL)
			new_gid = numeric(group, "group:");
		else
			new_gid = grp->gr_gid;
	}
}

static void
usage(void)
{
	switch (pers) {
	case PERS_CHOWN:
#ifdef	UCB
		eprintf(1, "usage: %s [-fhR] owner[.group] file ...\n",
				progname);
#else	/* !UCB */
		eprintf(1, "usage: %s [-h] [-R] uid file ...\n",
				progname);
#endif	/* !UCB */
		break;
	case PERS_CHGRP:
		eprintf(1, "%s: usage: %s [-h] [-R] gid file...\n",
				progname, progname);
		break;
	}
	exit(2);
}

int
main(int argc, char **argv)
{
#ifdef	UCB
	const char	optstring[] = "fhHLPR";
#else
	const char	optstring[] = "hHLPR";
#endif
	int i;

#ifdef	__GLIBC__
	putenv("POSIXLY_CORRECT=1");
#endif
	progname = basename(argv[0]);
	if (!strncmp(progname, "chgrp", 5))
		pers = PERS_CHGRP;
	while ((i = getopt(argc, argv, optstring)) != EOF) {
		switch (i) {
		case 'f':
			fflag = 1;
			break;
		case 'h':
			hflag = 1;
			break;
		case 'R':
			Rflag = 1;
			break;
		case 'H':
		case 'L':
		case 'P':
			HLPflag = i;
			break;
		default:
			usage();
		}
	}
	if (optind >= argc)
		usage();
	getowner(argv[optind++]);
	if (optind >= argc)
		usage();
	myeuid = geteuid();
	while (optind < argc)
		dochown(argv[optind++], 0);
	return errcnt;
}
